File: Python3 Solutions/F.py
class TrieNode:
    def __init__(self, char):
        self.letter = char
        self.children = [None] * 3
        self.is_end_of_word = False


def insert(root, word):
    cur = root
    for letter in word:
        index = ord(letter) - ord('a')
        if cur.children[index] == None:
            cur.children[index] = TrieNode(letter)
        cur = cur.children[index]
    
    cur.is_end_of_word = True


def check(query, root):
    for i, original_letter in enumerate(query):
        for ith in range(3):
            letter = chr(ith + ord('a'))
            if letter != query[i]:
                query[i] = letter
                found = search(root, query)
                if found:
                    return True
                query[i] = original_letter
    return False


def hasWord(cur, word_index, word, diff_count):
    if diff_count > 1:
        return False
    
    if (word_index < len(word) and cur.is_end_of_word):
        return False
        
    if (word_index >= len(word) and cur.is_end_of_word == False):
        return False
        
    if word_index == len(word) and cur.is_end_of_word == True and diff_count == 1:
        return True
    
    has_word = False
    for index in range(3):
        if cur.children[index] != None:
            add = 1 if (word_index < len(word) and word[word_index] != chr(index + ord('a'))) else 0
            found = hasWord(cur.children[index], word_index + 1, word, diff_count + add)
                
            if found:
                has_word = True
                break
                
    return has_word

    
def main():
    n, m = map(int, input().split(' '))
    root = TrieNode('*')

    for _ in range(n):
        word = input()
        insert(root, word)

    for _ in range(m):
        query = list(input())
        found = hasWord(root, 0, query, 0)
        if found:
            print("YES")
        else:
            print("NO")

if __name__ == "__main__":
    main()

File: Contests/ A2SV G123 Second Contest/F.cpp
#include <bits/stdc++.h>
using namespace std;

struct TrieNode {
	TrieNode* kids[3];
	bool isEnd = false;
};

void insert(TrieNode *root, string &word) {
	TrieNode *cur = root;
	for (char &c : word) {
		int index = c - 'a';
		if (cur->kids[index] == NULL) {
			cur->kids[index] = new TrieNode();
		}
		cur = cur->kids[index];
	}
	cur->isEnd = true;
}

bool dfs(TrieNode *cur, string &query, int &queryLen, int queryIdx, int diffCount) {

	if (diffCount > 1) {
		return false;
	}

	if (queryIdx == queryLen && cur->isEnd == false) {
		return false;
	}

	if (queryIdx < queryLen && cur->isEnd == true) {
		return false;
	}

	if (queryIdx == queryLen && cur->isEnd == true) {
		return diffCount == 1;
	}

	bool found = false;
	for (int i = 0; i <= 2; i++) {

		if (cur->kids[i] != NULL) {
			char curLetter = i + 'a';
			char queryLetter = query[queryIdx];
			bool validQueryIndex = queryIdx < queryLen;
			int diffAdd = (validQueryIndex && curLetter != queryLetter) ? 1 : 0;

			found = dfs(cur->kids[i], query, queryLen, queryIdx + 1, diffCount + diffAdd);
			if (found) {
				break;
			}
		}
	}

	return found;
}

int main() {
	
	int n, m;
	cin >> n >> m;

	TrieNode *root = new TrieNode();

	while (n--) {
		string word;
		cin >> word;
		insert(root, word);
	}
	
	while (m--) {
		string query;
		cin >> query;
		int queryLen = query.size();
		bool found = dfs(root, query, queryLen, 0, 0);
		if (found) {
			cout << "YES\n";
		}
		else {
			cout << "NO\n";
		}
	}

	return 0;
}